### DISCUSSION TITLE
Has anyone managed to get a TLS connection working with diesel_async & deadpool_postgres?

### DISCUSSION THREAD

**Discussion by NexRX** [weiznich/diesel_async#66] [Q&A] [answered]
I'm currently playing arround with some dependencies. Which right now involves me trying to get a Asynchronous, pooled, diesel connection to postgres. Thats quite a lot and at the moment I've got all of it wokring, except the TLS part. 

Every time I use TLS i get `CouldntSetupConfiguration(DatabaseError(UnableToSendCommand, "connection closed"))`

That is unless I make it fail in otherways: 
- I give the password wrong
- I Don't use TLS (TLS is required)
- Supply wrong cert

Anyone else got it to work? I would love to see how!

**Comment by weiznich** [comment, marked as answer]
See [this example](https://github.com/weiznich/diesel_async/pull/64). You need some additional setup for TLS. There are different crates out there depending on which TLS implementation you want to use.

**Reply by NexRX** [reply]
Ah hah, i was essentially doing that example but i didn't realise you have probe the TLS connection first yourself with 
```rust
    tokio::spawn(async move {
        if let Err(e) = connection.await {
            eprintln!("connection error: {}", e);
        }
    });
```
Not doing this is what caused me to error in the ways I did.

**Comment by liningpan** [comment]
I just want to add here, in case someone else try to do the same thing in the future. When the TLS connection is configured incorrectly, `bb8` will not surface the connection errors but will keep retrying and give a timeout error instead (issue https://github.com/djc/bb8/issues/141), with no explanation. This makes debugging nearly impossible. `deadpool` seems to be able to return the error properly. 

Also `tokio-postgres` doesn't recognize and will reject query parameters associated with mutual tls like `sslrootcert`, `sslcert`, and `sslkey`, so you have to pass necessary information to the setup function in a different way. For example, you can use environment variables or make the `establish_connection` function a closure that captures some additional tls config.

https://github.com/weiznich/diesel_async/blob/67c677e5be5e31f0d8d7d333cd31a81fa50d585f/examples/postgres/pooled-with-rustls/src/main.rs#L15-L18

Related issues:
https://github.com/jbg/tokio-postgres-rustls/issues/11
https://github.com/sfackler/rust-postgres/issues/768

**Reply by weiznich** [reply]
Thanks for adding this comment here. Mind submitting a PR that adds the information about `sslcrt` etc to the example linked above? I think that would help future users to implement such a setup more easily.
